packages/api/src/codegen/typescript.ts

```diff
@@ -0,0 +1,357 @@
+      if (data.types.params) {
+        if (data.types.params.body || data.types.params.formData) {
+          // @todo this should be an optional param if there's no required params.
+          parameters.push({ name: 'body', type: data.types.params.body });
+          fetchArgs.push('body');
+        }
+
+        if (data.types.params.metadata) {
+          // @todo this should be an optional param if there's no required params.
+          parameters.push({ name: 'metadata', type: data.types.params.metadata });
+          fetchArgs.push('metadata');
+        }
+      }
+
+      let returnType: string;
+      if (data.types.responses) {
+        returnType = `Promise<${Object.values(data.types.responses).join(' | ')}>`;
+      }
+
+      // @todo add docblock tags for these params and return types
+      const method = sdk.addMethod({
+        name: operationId,
+        returnType,
+        parameters,
+        statements: `return this.core.fetch(${fetchArgs.join(', ')});`,
+      });
+
+      let docblock;
+      const summary = operation.getSummary();
+      const description = operation.getDescription();
+      if (summary || description) {
+        docblock = [
+          summary ? `${summary.replace(/\n/g, '\n * ')}` : '',
+          description ? `${description.replace(/\n/g, '\n * ')}` : '',
+        ]
+          .filter(Boolean)
+          .join('\n\n');
+
+        method.addJsDoc(docblock);
+      }
+
+      // Add a typed generic HTTP method overload for this operation.
+      if (operation.method in methodGenerics) {
+        methodGenerics[operation.method].addOverload({
+          parameters: [{ name: 'path', type: 'string' }, ...parameters],
+          returnType,
+          docs: docblock ? [docblock] : null,
+        });
+      }
+    });
+
+    // @todo should all of these isolated into their own file outside of the main sdk class file?
+    // Add all known types that we're using into the SDK.
+    Array.from(this.types.values()).forEach(exp => {
+      sdkSource.addStatements(exp);
+    });
+
+    // const result = project.emitToMemory();
+    return this.project.getSourceFiles().map(sourceFile => ({
+      [sourceFile.getBaseName()]: sourceFile.getFullText(),
+    }));
+  }
+
+  /**
+   * Convert a JSON Schema object into a readily available TypeScript type or interface along with
+   * any `$ref` pointers that are in use and turn those into TS types too.
+   *
+   * Under the hood this uses https://npm.im/json-schema-to-typescript for all composition and
+   * conversion.
+   *
+   * @param schema
+   * @param name
+   */
+  async convertJSONSchemaToTypescript(schema: JSONSchema, name: string) {
+    // Though our JSON Schema type exposes JSONSchema4, which `json-schema-to-typescript`, it won't
+    // accept our custom type.
+    const ts = await compile(schema as any, name, {
+      bannerComment: '',
+    });
+
+    let primaryType: string;
+    const tmp = this.project.createSourceFile(`${name}.types.tmp.ts`, ts);
```

Comment:
Because [json-schema-to-typescript](https://npm.im/json-schema-to-typescript) outputs only a string containing our exported TS types we need to do a bit of post-processing on it in order to extract a few things:

* The primary type that this JSON Schema object was created for (as dictated by the `name` argument). So if we're converting the `metadata` parameter JSON Schema for a `findByPets` operation, and `name` in this case is `FindByPetsMetadata`, then we'll be looking the primary type of `FindByPetsMetadata` so we can add it into the `findByPets()` operation accessor.
* All types and interfaces that were created with this JSON Schema object. Because we'll need to actually insert these types and interfaces into our generated SDK we need to make sure that we don't duplicate them so we need to do some more post-processing here to put them into a set of the exported name to the exported TS type string content.

If this all sounds like a lot to mentally process it's because it is.

This code also isn't optimized either because we're going to definitely be processing the same JSON Schema object sometimes. A problem for another day.